CONFIG_DEVICE_DT_METADATA and Software-Device does not compile

**Describe the bug**
When the kconfig option CONFIG_DEVICE_DT_METADATA is enabled (which gets enabled by CONFIG_GPIO_SHELL), compilation fails if a software device (DT_INVALID_NODE) is present.  
In my specific scenario, this issue arises on one board due to the PPP device, and on another board due to the three additional LAN ports generated by the DSA driver.

**To Reproduce**
It can be reproduced by
`west build -p -b b_u585i_iot02a zephyr/samples/net/cellular_modem/ -DCONFIG_DEVICE_DT_METADATA=y`
or
`west build -p -b ip_k66f zephyr/samples/net/dsa/ -DCONFIG_DEVICE_DT_METADATA=y`

**Impact**
No big impact, i just use `CONFIG_GPIO_SHELL=y` and `CONFIG_DEVICE_DT_METADATA=n`

**Logs and console output**
Logs are quite long, i shorted them as they same output gets repeated multiple times.
cellular_modem:
```

FAILED: zephyr/drivers/modem/CMakeFiles/drivers__modem.dir/modem_cellular.c.obj 
ccache /opt/toolchains/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DHSE_VALUE=8000000 -DKERNEL -DK_HEAP_MEM_POOL_SIZE=1024 -DPICOLIBC_LONG_LONG_PRINTF_SCANF -DSTM32U585xx -DUSE_FULL_LL_DRIVER -DUSE_HAL_DRIVER -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/workdir/zephyr/subsys/net/ip -I/workdir/zephyr/subsys/net/lib/sockets -I/workdir/build/zephyr/include/generated/zephyr -I/workdir/zephyr/include -I/workdir/build/zephyr/include/generated -I/workdir/zephyr/soc/st/stm32 -I/workdir/zephyr/include/zephyr/posix -I/workdir/zephyr/lib/posix/options/getopt -I/workdir/zephyr/soc/st/stm32/common/. -I/workdir/zephyr/drivers -I/workdir/zephyr/soc/st/stm32/stm32u5x/. -I/workdir/zephyr/subsys/net/lib/dns/. -I/workdir/zephyr/subsys/net/conn_mgr/. -I/workdir/modules/hal/cmsis/CMSIS/Core/Include -I/workdir/zephyr/modules/cmsis/. -I/workdir/modules/hal/stm32/stm32cube/stm32u5xx/soc -I/workdir/modules/hal/stm32/stm32cube/stm32u5xx/drivers/include -I/workdir/modules/hal/stm32/stm32cube/common_ll/include -isystem /workdir/zephyr/lib/libc/common/include -Wshadow -fno-strict-aliasing -Os -imacros /workdir/build/zephyr/include/generated/zephyr/autoconf.h -fno-printf-return-value -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m33 -mthumb -mabi=aapcs -mfp16-format=ieee -mtp=soft --sysroot=/opt/toolchains/zephyr-sdk-0.16.1/arm-zephyr-eabi/arm-zephyr-eabi -imacros /workdir/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/workdir/zephyr/samples/net/cellular_modem=CMAKE_SOURCE_DIR -fmacro-prefix-map=/workdir/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/workdir=WEST_TOPDIR -ffunction-sections -fdata-sections --specs=picolibc.specs -D_POSIX_THREADS -std=c99 -MD -MT zephyr/drivers/modem/CMakeFiles/drivers__modem.dir/modem_cellular.c.obj -MF zephyr/drivers/modem/CMakeFiles/drivers__modem.dir/modem_cellular.c.obj.d -o zephyr/drivers/modem/CMakeFiles/drivers__modem.dir/modem_cellular.c.obj -c /workdir/zephyr/drivers/modem/modem_cellular.c
In file included from /workdir/zephyr/include/zephyr/sys/util_macro.h:34,
                 from /workdir/zephyr/include/zephyr/sys/util.h:17,
                 from /workdir/zephyr/include/zephyr/sys/atomic.h:18,
                 from /workdir/zephyr/include/zephyr/kernel_includes.h:25,
                 from /workdir/zephyr/include/zephyr/kernel.h:17,
                 from /workdir/zephyr/drivers/modem/modem_cellular.c:7:
/workdir/zephyr/include/zephyr/devicetree.h:87:25: error: '__NODELABEL_NUM' undeclared here (not in a function); did you mean 'DT_N_NODELABEL_NUM'?
   87 | #define DT_INVALID_NODE _
      |                         ^
/workdir/zephyr/include/zephyr/sys/util_internal.h:72:26: note: in definition of macro '__DEBRACKET'
   72 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_macro.h:180:9: note: in expansion of macro 'Z_COND_CODE_1'
  180 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/devicetree.h:4528:9: note: in expansion of macro 'COND_CODE_1'
 4528 |         COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT),   \
      |         ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_internal.h:69:53: note: in expansion of macro '__DEBRACKET'
   69 | #define __GET_ARG2_DEBRACKET(ignore_this, val, ...) __DEBRACKET val
      |                                                     ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_macro.h:180:9: note: in expansion of macro 'Z_COND_CODE_1'
  180 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_macro.h:224:9: note: in expansion of macro 'COND_CODE_1'
  224 |         COND_CODE_1(_flag, _code, ())
      |         ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/device.h:1209:9: note: in expansion of macro 'IF_ENABLED'
 1209 |         IF_ENABLED(CONFIG_DEVICE_DT_METADATA,                                   \
      |         ^~~~~~~~~~
/workdir/zephyr/include/zephyr/devicetree.h:2401:36: note: in expansion of macro 'DT_CAT'
 2401 | #define DT_NUM_NODELABELS(node_id) DT_CAT(node_id, _NODELABEL_NUM)
      |                                    ^~~~~~
/workdir/zephyr/include/zephyr/device.h:1020:35: note: in expansion of macro 'DT_NUM_NODELABELS'
 1020 |                 .num_nodelabels = DT_NUM_NODELABELS(node_id),           \
      |                                   ^~~~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/device.h:1210:21: note: in expansion of macro 'Z_DEVICE_DT_METADATA_DEFINE'
 1210 |                    (Z_DEVICE_DT_METADATA_DEFINE(node_id, dev_id);))             \
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/net/net_if.h:3194:9: note: in expansion of macro 'Z_DEVICE_DEFINE'
 3194 |         Z_DEVICE_DEFINE(node_id, dev_id, name, init_fn, pm, data,       \
      |         ^~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/net/net_if.h:3202:9: note: in expansion of macro 'Z_NET_DEVICE_INIT_INSTANCE'
 3202 |         Z_NET_DEVICE_INIT_INSTANCE(node_id, dev_id, name, 0, init_fn,   \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/net/net_if.h:3227:9: note: in expansion of macro 'Z_NET_DEVICE_INIT'
 3227 |         Z_NET_DEVICE_INIT(DT_INVALID_NODE, dev_id, name, init_fn, pm,   \
      |         ^~~~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/net/net_if.h:3227:27: note: in expansion of macro 'DT_INVALID_NODE'
 3227 |         Z_NET_DEVICE_INIT(DT_INVALID_NODE, dev_id, name, init_fn, pm,   \
      |                           ^~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/modem/ppp.h:192:9: note: in expansion of macro 'NET_DEVICE_INIT'
  192 |         NET_DEVICE_INIT(_CONCAT(ppp_net_dev_, _name), "modem_ppp_" # _name,                        \
      |         ^~~~~~~~~~~~~~~
/workdir/zephyr/drivers/modem/modem_cellular.c:2113:9: note: in expansion of macro 'MODEM_PPP_DEFINE'
 2113 |         MODEM_PPP_DEFINE(MODEM_CELLULAR_INST_NAME(ppp, inst), NULL, 98, 1500, 64);                 \
      |         ^~~~~~~~~~~~~~~~
/workdir/build/zephyr/include/generated/zephyr/devicetree_generated.h:26026:47: note: in expansion of macro 'MODEM_CELLULAR_DEVICE_QUECTEL_BG95'
26026 | #define DT_FOREACH_OKAY_INST_quectel_bg95(fn) fn(0)
      |                                               ^~
/workdir/zephyr/include/zephyr/sys/util_internal.h:105:36: note: in expansion of macro 'DT_FOREACH_OKAY_INST_quectel_bg95'
  105 | #define UTIL_PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
      |                                    ^
/workdir/zephyr/drivers/modem/modem_cellular.c:2618:1: note: in expansion of macro 'DT_INST_FOREACH_STATUS_OKAY'
 2618 | DT_INST_FOREACH_STATUS_OKAY(MODEM_CELLULAR_DEVICE_QUECTEL_BG95)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Dsa:
```
In file included from /workdir/zephyr/include/zephyr/sys/util_macro.h:34,
                 from /workdir/zephyr/include/zephyr/sys/util.h:17,
                 from /workdir/zephyr/include/zephyr/sys/cbprintf_internal.h:15,
                 from /workdir/zephyr/include/zephyr/sys/cbprintf.h:124,
                 from /workdir/zephyr/include/zephyr/logging/log_msg.h:11,
                 from /workdir/zephyr/include/zephyr/logging/log_core.h:9,
                 from /workdir/zephyr/include/zephyr/logging/log.h:11,
                 from /workdir/zephyr/drivers/ethernet/dsa_ksz8xxx.c:9:
/workdir/zephyr/include/zephyr/devicetree.h:87:25: error: '__NODELABEL_NUM' undeclared here (not in a function); did you mean 'DT_N_NODELABEL_NUM'?
   87 | #define DT_INVALID_NODE _
      |                         ^
/workdir/zephyr/include/zephyr/sys/util_internal.h:72:26: note: in definition of macro '__DEBRACKET'
   72 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_macro.h:180:9: note: in expansion of macro 'Z_COND_CODE_1'
  180 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/devicetree.h:4528:9: note: in expansion of macro 'COND_CODE_1'
 4528 |         COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT),   \
      |         ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_internal.h:69:53: note: in expansion of macro '__DEBRACKET'
   69 | #define __GET_ARG2_DEBRACKET(ignore_this, val, ...) __DEBRACKET val
      |                                                     ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_macro.h:180:9: note: in expansion of macro 'Z_COND_CODE_1'
  180 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/sys/util_macro.h:224:9: note: in expansion of macro 'COND_CODE_1'
  224 |         COND_CODE_1(_flag, _code, ())
      |         ^~~~~~~~~~~
/workdir/zephyr/include/zephyr/device.h:1209:9: note: in expansion of macro 'IF_ENABLED'
 1209 |         IF_ENABLED(CONFIG_DEVICE_DT_METADATA,                                   \
      |         ^~~~~~~~~~
/workdir/zephyr/include/zephyr/devicetree.h:2401:36: note: in expansion of macro 'DT_CAT'
 2401 | #define DT_NUM_NODELABELS(node_id) DT_CAT(node_id, _NODELABEL_NUM)
      |                                    ^~~~~~
/workdir/zephyr/include/zephyr/device.h:1020:35: note: in expansion of macro 'DT_NUM_NODELABELS'
 1020 |                 .num_nodelabels = DT_NUM_NODELABELS(node_id),           \
      |                                   ^~~~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/device.h:1210:21: note: in expansion of macro 'Z_DEVICE_DT_METADATA_DEFINE'
 1210 |                    (Z_DEVICE_DT_METADATA_DEFINE(node_id, dev_id);))             \
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/net/net_if.h:3180:9: note: in expansion of macro 'Z_DEVICE_DEFINE'
 3180 |         Z_DEVICE_DEFINE(node_id, dev_id, name, init_fn, pm, data,       \
      |         ^~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/net/net_if.h:3277:9: note: in expansion of macro 'Z_NET_DEVICE_INIT_INSTANCE'
 3277 |         Z_NET_DEVICE_INIT_INSTANCE(DT_INVALID_NODE, dev_id, name,       \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
/workdir/zephyr/include/zephyr/net/net_if.h:3277:36: note: in expansion of macro 'DT_INVALID_NODE'
 3277 |         Z_NET_DEVICE_INIT_INSTANCE(DT_INVALID_NODE, dev_id, name,       \
      |                                    ^~~~~~~~~~~~~~~
/workdir/zephyr/drivers/ethernet/dsa_ksz8xxx.c:1122:9: note: in expansion of macro 'NET_DEVICE_INIT_INSTANCE'
 1122 |         NET_DEVICE_INIT_INSTANCE(CONCAT(dsa_slave_port_, slave),           \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
/workdir/build/zephyr/include/generated/zephyr/devicetree_generated.h:12016:72: note: in expansion of macro 'NET_SLAVE_DEVICE_INIT_INSTANCE'
12016 | #define DT_N_S_soc_S_spi_4002d000_S_dsa_0_FOREACH_CHILD_VARGS(fn, ...) fn(DT_N_S_soc_S_spi_4002d000_S_dsa_0_S_lan_3, __VA_ARGS__) fn(DT_N_S_soc_S_spi_4002d000_S_dsa_0_S_lan_2, __VA_ARGS__) fn(DT_N_S_soc_S_spi_4002d000_S_dsa_0_S_lan_1, __VA_ARGS__)
      |                                                                        ^~
/workdir/zephyr/include/zephyr/devicetree.h:4781:24: note: in expansion of macro 'DT_N_S_soc_S_spi_4002d000_S_dsa_0_FOREACH_CHILD_VARGS'
 4781 | #define DT_CAT(a1, a2) a1 ## a2
      |                        ^~
/workdir/zephyr/include/zephyr/devicetree.h:3728:9: note: in expansion of macro 'DT_FOREACH_CHILD_VARGS'
 3728 |         DT_FOREACH_CHILD_VARGS(DT_DRV_INST(inst), fn, __VA_ARGS__)
      |         ^~~~~~~~~~~~~~~~~~~~~~
/workdir/zephyr/drivers/ethernet/dsa_ksz8xxx.c:1166:9: note: in expansion of macro 'DT_INST_FOREACH_CHILD_VARGS'
 1166 |         DT_INST_FOREACH_CHILD_VARGS(n, NET_SLAVE_DEVICE_INIT_INSTANCE, n);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/workdir/build/zephyr/include/generated/zephyr/devicetree_generated.h:12622:52: note: in expansion of macro 'DSA_DEVICE'
12622 | #define DT_FOREACH_OKAY_INST_microchip_ksz8794(fn) fn(0)
      |                                                    ^~
/workdir/zephyr/include/zephyr/sys/util_internal.h:105:36: note: in expansion of macro 'DT_FOREACH_OKAY_INST_microchip_ksz8794'
  105 | #define UTIL_PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
      |                                    ^
/workdir/zephyr/drivers/ethernet/dsa_ksz8xxx.c:1169:1: note: in expansion of macro 'DT_INST_FOREACH_STATUS_OKAY'
 1169 | DT_INST_FOREACH_STATUS_OKAY(DSA_DEVICE);
```


**Environment:**
Should not matter but just in case
 - OS: Linux
 - Toolchain:  Zephyr SDK (docker.io/zephyrprojectrtos/zephyr-build:v0.26.13)
 - Commit SHA: 7f0f0a4f97d6adb6c134db0788dd200fbfba8515 and 18ebfe59d043f8b53bdd0c7ed48e2c9920abb881

--- a/include/zephyr/device.h
+++ b/include/zephyr/device.h
@@ -1019,7 +1019,7 @@ device_get_dt_nodelabels(const struct device *dev)
 		.state = (state_),							\
 		.data = (data_),							\
 		IF_ENABLED(CONFIG_DEVICE_DEPS, (.deps = (deps_),)) /**/			\
-		IF_ENABLED(CONFIG_PM_DEVICE, ({ .pm_base = (pm_),})) /**/		\
+		IF_ENABLED(CONFIG_PM_DEVICE, ({ .pm_base = (pm_),},)) /**/		\
 		IF_ENABLED(CONFIG_DEVICE_DT_METADATA,					\
 			   (.dt_meta = &Z_DEVICE_DT_METADATA_NAME_GET(dev_id_),))	\
 	}

--- a/include/zephyr/device.h
+++ b/include/zephyr/device.h
@@ -1008,10 +1008,11 @@ device_get_dt_nodelabels(const struct device *dev)
  * @param api_ Reference to device API ops.
  * @param state_ Reference to device state.
  * @param deps_ Reference to device dependencies.
+ * @param node_id_ Devicetree node identifier
  * @param dev_id_ Device identifier token, as passed to Z_DEVICE_BASE_DEFINE
  */
-#define Z_DEVICE_INIT(name_, pm_, data_, config_, api_, state_, deps_,			\
-			dev_id_)			\
+#define Z_DEVICE_INIT(name_, pm_, data_, config_, api_, state_, deps_, node_id_,	\
+		      dev_id_)								\
 	{										\
 		.name = name_,								\
 		.config = (config_),							\
@@ -1021,7 +1022,9 @@ device_get_dt_nodelabels(const struct device *dev)
 		IF_ENABLED(CONFIG_DEVICE_DEPS, (.deps = (deps_),)) /**/			\
 		IF_ENABLED(CONFIG_PM_DEVICE, ({ .pm_base = (pm_),},)) /**/		\
 		IF_ENABLED(CONFIG_DEVICE_DT_METADATA,					\
-			   (.dt_meta = &Z_DEVICE_DT_METADATA_NAME_GET(dev_id_),))	\
+			   (IF_ENABLED(DT_NODE_EXISTS(node_id_),			\
+				       (.dt_meta = &Z_DEVICE_DT_METADATA_NAME_GET(	\
+						dev_id_),))))				\
 	}
 
 /**
@@ -1056,7 +1059,7 @@ device_get_dt_nodelabels(const struct device *dev)
 	STRUCT_SECTION_ITERABLE_NAMED_ALTERNATE(                                                   \
 		device, COND_CODE_1(Z_DEVICE_IS_MUTABLE(node_id), (device_mutable), (device)),     \
 		Z_DEVICE_SECTION_NAME(level, prio), DEVICE_NAME_GET(dev_id)) =                     \
-		Z_DEVICE_INIT(name, pm, data, config, api, state, deps, dev_id)
+		Z_DEVICE_INIT(name, pm, data, config, api, state, deps, node_id, dev_id)
 
 /* deprecated device initialization levels */
 #define Z_DEVICE_LEVEL_DEPRECATED_EARLY                                        \
@@ -1142,7 +1145,8 @@ device_get_dt_nodelabels(const struct device *dev)
 		   (Z_DEVICE_DEPS_DEFINE(node_id, dev_id, __VA_ARGS__);))       \
                                                                                 \
 	IF_ENABLED(CONFIG_DEVICE_DT_METADATA,                                   \
-		   (Z_DEVICE_DT_METADATA_DEFINE(node_id, dev_id);))             \
+		   (IF_ENABLED(DT_NODE_EXISTS(node_id),                         \
+			      (Z_DEVICE_DT_METADATA_DEFINE(node_id, dev_id);))))\
                                                                                 \
 	Z_DEVICE_BASE_DEFINE(node_id, dev_id, name, pm, data, config, level,    \
 		prio, api, state, Z_DEVICE_DEPS_NAME(dev_id));                   \

